ion-schema/src/nfa.rs

```diff
@@ -0,0 +1,258 @@
+                visits.insert(0, 1);
+                visits
+            },
+            nfa,
+        }
+    }
+
+    pub fn has_final_state(&self) -> bool {
+        self.visits
+            .iter()
+            .any(|(id, _)| self.nfa.states[*id] == State::Final)
+    }
+
+    pub fn transition(&mut self, event: Event, type_store: &TypeStore) -> HashMap<usize, usize> {
+        let mut new_states = HashMap::new();
+        for (source_state_id, visits) in self.visits.iter() {
+            let can_exit = self.nfa.states[*source_state_id].allows_exit_after_n_visits(*visits);
+            let transitions: &HashSet<Transition> =
+                self.nfa.transitions.get(source_state_id).unwrap();
+            self.evaluate_transitions(
+                *source_state_id,
+                transitions,
+                can_exit,
+                *visits,
+                event.to_owned(),
+                type_store,
+                &mut new_states,
+            );
+        }
+        new_states
+    }
+
+    fn evaluate_transitions(
+        &self,
+        source_state_id: usize,
+        transitions: &HashSet<Transition>,
+        can_exit: bool,
+        visits: usize,
+        event: Event,
+        type_store: &TypeStore,
+        new_states: &mut HashMap<usize, usize>,
+    ) {
+        let mut destination_states_for_optional_state: HashSet<Transition> = HashSet::new();
+        for transition in transitions {
+            let destination_state_id = transition.destination;
+            let destination_state = &self.nfa.states[destination_state_id];
+
+            if destination_state_id == source_state_id {
+                if self.nfa.states[destination_state_id].allows_n_visits(visits + 1) {
+                    let can_enter = destination_state.allows_to_enter_for_ion_value(
+                        event.to_owned(),
+                        transition.type_id,
+                        type_store,
+                    );
+                    if can_enter {
+                        new_states.insert(destination_state_id, visits + 1);
+                    }
+                }
+            } else if can_exit {
+                let can_enter = destination_state.allows_to_enter_for_ion_value(
+                    event.to_owned(),
+                    transition.type_id,
+                    type_store,
+                );
+                if can_enter {
+                    new_states.insert(destination_state_id, 1);
+                }
+            }
+
+            // for optional state add transitions to next states skipping the optional state
+            if transition.to_optional_state {
+                let mut transitions = self
+                    .nfa
+                    .transitions
+                    .get(&destination_state_id)
+                    .unwrap()
+                    .to_owned();
+
+                // skip the optional state itself
+                transitions.remove(transition);
```

Comment:
This looks like a code smell to me. NFA graph should be in a usable form when you build it, and shouldn't require any modification when evaluating some input against the NFA. Can you explain why this is necessary?

Reply:
I am using this to simp0lify building NFA. The NFa that is built also contains transitions with `0` occurrences. Meaning for optional state I will have to traverse through all possible states and them in the Visit-count map that is created.
Example,
```
 NFA:
                                 +------------ 0 -------------+
                                 |                            |
                                 |                            V
     I(INITIAL) ----> S1(INTERMEDIATE(0, 1)) -- 1 --> S2(INTERMEDIATE(1, 1)) -- 1 --> S3(INTERMEDIATE(1, 1)) ----> F(FINAL)
```